### ISSUE TITLE
Preprocess batches and load those (Instead of processing on the fly)

### ISSUE THREAD

**Issue by jacobbieker** [openclimatefix/satflow#63]
The webdataset works decently well, but it still takes awhile for the timesteps to be loaded, cropped, normalized, etc. Prepreparing some cutouts could be quite useful, and make the dataloading process a lot quicker. Currently, nearly all my experiments use the last half hour of images to predict the next 1-4 hours, so just doing the processing offline would probably work better. Webdataset actually recommend that too, and recommends AIStore for ETL on clusters, etc. Not there yet in terms of need, but this could lead to a lot faster loading. Also can then be sure validation set is identical everytime.

**Comment by JackKelly** [comment]
Yeah, I would definitely recommend this ;) 

I was resistant to this idea until very recently because I thought it'd be faster play with different pre-processing tricks if the pre-processing was done on-the-fly but, after trying for literally _months_ to get the pre-processing fast enough (!), I've just started pre-processing entire batches and saving them to disk.  The end-result is 12x speedup in training :), and I think the bottleneck is now GPU FLOPS instead of disk speed.   Notes are here: https://github.com/openclimatefix/nowcasting_dataset/issues/58

I now save complete batches to disk as NetCDF4 files (compressed using LZF), one batch per NetCDF5 file. All the PyTorch Dataset has to do is grab a NetCDF file from disk and turn it into Tensors :), and I run 20 worker processes.  Which I'm guessing is very similar in spirit to your approach using webdataset (except I'm using NetCDF4 instead of numpy arrays, but NetCDF4 is just numpy arrays with human-readable labels for the dimensions and coordinates, which is useful for me because my memory is so bad I can't remember what went into which dimension!)

**Comment by JackKelly** [comment]
FWIW, I still think Zarr has a role to play, perhaps as an intermediate step.  So perhaps the data processing pipeline looks something like:

1. Raw data
2. Zarr with everything that was in the raw data (and perhaps re-projected)
3. Lots of copies from the Zarr data for different uses, in whatever format makes most sense (e.g. NetCDF or numpy arrays for different ML experiments; maybe JPEGs for the web UI front-end, etc.)

**Comment by jacobbieker** [comment]
Yeah, it seems like Zarr is a nice store to more easily make whatever copy of the data we want. And that pipeline is what I was thinking too! From previous experience, storage is usually a lot cheaper than paying for the computation over and over again to process the data, especially if that processing takes awhile to do, and doubly so if we are using the same data lots of times for training models, etc.

**Comment by jacobbieker** [comment]
Note: for 12 previous timesteps + 48 future timesteps,  128 pixels total size of a single example is 56mb, goes down to 46mb for 6 previous + 48 future, and 26.5mb for 6 previous + 24 future. So to not run out of space on internal SSD, preprocessing 30000 training, and 10000 validation examples.

**Comment by JackKelly** [comment]
Just curious: Are you compressing the data on disk?  (I must admit I've never saved numpy arrays to disk; so I don't know what's possible!)

**Comment by jacobbieker** [comment]
That was actually with saving the Torch tensors directly to disk with ```torch.save```, which I admittedly thought had some compression, but doesn't seem to! You can save compressed numpy arrays with ```np.savez_compressed``` which I'll see if that helps at all

**Comment by jacobbieker** [comment]
Yeah, using np.savez_compressed reduces the filesize of the last 6 + 24 future to 4.6mb, so switching to that

**Comment by jacobbieker** [comment]
Have been testing it training two models. For the CloudGAN, its not making it faster at all, probably as a result of CloudGAN's training steps just taking a long time to run. For a simple ConvLSTM, though, its a slight improvement, but not by much, up to 2it/s from 1-1.5. Where it does save is RAM usage, which instead of hovering around 60GB is at around 15GB. But overall, using npz files and loading them doesn't seem blazingly faster on my local machine than before. It could be that loading the spatial extant and just sample from the same area ~50 times with random crops and temporal starts makes up for the processing time required. Could also try just saving each batch to webdataset to see if having it all in a compressed tar file helps at all.

**Comment by JackKelly** [comment]
Heh, we're going though a surprisingly similar journeys!

On Thurs, I was all excited because saving batches to disk allowed my simple CNN to train at 20 it/s.

But, now I'm training The Perceiver, we're back to 2 it/s (bottlenecked by the GPU!)  Which is about the speed my load-on-the-fly dataloader managed!

(But, if we get nvidia A6000s; and/or faster GPUs in the cloud, then the bottleneck might revert to being the DataLoader!)

BTW, are you using mixed-precision training (setting `pl.Trainer(precision=16)`?)  That seems to help a little (not much though).  (Sorry for being too lazy to open a new issue!)

**Comment by jacobbieker** [comment]
With mixed precision, the ConvLSTM gives 3.62it/s so it is nearly twice as fast, but yeah, I think the GPU is a bigger bottleneck than I thought!

**Comment by jacobbieker** [comment]
Another example, with the Skillful Nowcasting GAN, it takes 11-12s/it on my laptop to run in training, but runs at 1.8-2.8it/s in validation mode.
